### PR TITLE
feat(ui,api): alternate workflow list UX

### PR DESCRIPTION
**Pull request by maryhipp**
## Summary

Swaps out our workflow library modal for a list within the left-hand panel similar to prompt templates.

## QA Instructions

Screenshot 2024-10-07 at 9 09 55 PM
Screenshot 2024-10-07 at 9 10 06 PM
Screenshot 2024-10-07 at 9 10 16 PM
Screenshot 2024-10-07 at 9 10 23 PM

## Checklist

- [ ] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _Documentation added / updated (if applicable)_

### CODE REVIEW

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
+const overlayScrollbarsStyles: CSSProperties = {
+  height: '100%',
+  width: '100%',
+};
+
+export const selectCleanEditor = createMemoizedSelector([selectNodesSlice, selectWorkflowSlice], (nodes, workflow) => {
+  const noNodes = !nodes.nodes.length;
+  const isTouched = workflow.isTouched;
+  return noNodes && !isTouched;
+});
```
Comment (maryhipp) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:35]: My best guess at how to assume that a workflow editor is in a truly clean state - no nodes, no changes

Comment (psychedelicious) [reply invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:35]: Seems reasonable to me

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/SaveWorkflowButton.tsx [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import { memo, useCallback } from 'react';
+import { useTranslation } from 'react-i18next';
+import { PiFloppyDiskBold } from 'react-icons/pi';
+
+const SaveWorkflowButton = () => {
+  const { t } = useTranslation();
+  const { onOpen } = useSaveWorkflowAsDialog();
+  const { saveWorkflow } = useSaveLibraryWorkflow();
+
+  const handleClickSave = useCallback<MouseEventHandler<HTMLButtonElement>>(
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/SaveWorkflowButton.tsx:15]: If you click the save button in view mode, the dialog doesn't show. Then if you click edit, it pops up. Looks like the dialog component is only rendered in the edit view

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/WorkflowList.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        </Flex>
+      </Button>
+      <Collapse in={isOpen}>
+        {data.length ? (
+          data.map((workflow) => <WorkflowListItem workflow={workflow} key={workflow.workflow_id} />)
+        ) : (
+          <IAINoContentFallback
+            fontSize="sm"
+            py={4}
+            label={searchTerm ? t('stylePresets.noMatchingTemplates') : t('stylePresets.noTemplates')}
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/WorkflowList.tsx:35]: maybe replace "templates" w/ "workflows"

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/WorkflowListItem.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                    </Tooltip>
+                    {isActive && (
+                      <Badge
+                        color="invokeBlue.400"
+                        borderColor="invokeBlue.700"
+                        borderWidth={1}
+                        bg="transparent"
+                        flexShrink={0}
+                      >
+                        {t('stylePresets.active')}
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/WorkflowListMenu/WorkflowListItem.tsx]: maybe "selected" or "opened" or something like that is more appropriate than "active"

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx [hunk 5]
```diff
@@ -1,26 +1,51 @@
...
+import { WorkflowListMenuTrigger } from './WorkflowListMenu/WorkflowListMenuTrigger';
 
 const panelGroupStyles: CSSProperties = { height: '100%', width: '100%' };
 
+const overlayScrollbarsStyles: CSSProperties = {
+  height: '100%',
+  width: '100%',
+};
+
+export const selectCleanEditor = createMemoizedSelector([selectNodesSlice, selectWorkflowSlice], (nodes, workflow) => {
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:31]: When the selector returns a primitive data type, you can use the default `createSelector` - no need for the memoized one. not gonna hurt anything tho

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx (hunk 1 above)
Comment (psychedelicious) [reply invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:35]: Maybe add `workflow.id === undefined`? IIRC that means its an unsaved workflow

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx [hunk 6]
```diff
@@ -1,26 +1,51 @@
...
   const panelGroupRef = useRef<ImperativePanelGroupHandle>(null);
+  const isWorkflowListMenuOpen = useStore($isWorkflowListMenuIsOpen);
+
+  const isCleanEditor = useAppSelector(selectCleanEditor);
+
+  useEffect(() => {
+    if (isCleanEditor) {
+      $isWorkflowListMenuIsOpen.set(true);
+    }
+  }, [isCleanEditor]);
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:48]: There is an awkward sequence:
- Editing a workflow
- Delete all nodes
- Save it
- The menu opens bc `isCleanEditor === true`

Maybe the suggested change the selector would resolve this

**Review by psychedelicious — APPROVED** [review]
Couple small things to tidy but otherwise LGTM

I tweaked the layout on the list item

Once the comments are addressed to your satisfaction feel free to merge

**Comment by psychedelicious** [comment]
Looks like there is a circular dep, for context we are checking for circular deps primarily bc it can break hot reload. But also it can cause some problems during builds. Sometimes its a pain to resolve sorry.

File: invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx (hunk 6 above)
Comment (maryhipp) [reply invokeai/frontend/web/src/features/nodes/components/sidePanel/NodeEditorPanelGroup.tsx:48]: Yep that seems to have fixed it

**Pull request merged**